osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -18,146 +23,318 @@
-            final DbEncryptionKeyReference keyReference = secret.getEncryptionKeyReference();
-            if (keyReference == null) {
-                throw new IllegalStateException("Could not create encrypted secret: secret has no key reference");
-            }
-            final byte[] secretBytes = HexUtils.fromHexString(secret.getEncodedSecret());
-            final EncryptedSecret encryptedSecret = new EncryptedSecret(keyReference.getEncryptionProviderType(),
-                    secretBytes);
-            return this.createTypedSecret(secret, keyReference, encryptedSecret);
-        } else {    //Should never happen because of stream mapping in retrieveSecrets()
-            throw new IllegalStateException("Could not create typed secret for NULL secret");
+    public synchronized void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) {
+        secrets.forEach(s -> this.checkNrNewSecretsOfType(deviceIdentification, s.getSecretType(), 0));
+        List<EncryptedTypedSecret> aesSecrets = secrets.stream().map(ts -> new EncryptedTypedSecret(ts.getSecret(),
+                ts.getSecretType())).map(this::reencryptRsa2Aes).collect(toList());
+        this.storeAesSecrets(deviceIdentification, aesSecrets);
+    }
+
+    private void storeAesSecrets(final String deviceIdentification, final List<EncryptedTypedSecret> secrets) {
+        secrets.stream().map(this::validateAndReturnNewSecret).map(ets -> this
+                .createDbEncrypted(deviceIdentification, ets, this.getKeyByReference(ets.encryptionKeyReference)))
+               .collect(collectingAndThen(toList(), this.secretRepository::saveAll));
+    }
+
+    public synchronized void activateNewSecrets(final String deviceIdentification, final List<SecretType> secretTypes) {
+        secretTypes.stream().forEach(t -> this.checkNrNewSecretsOfType(deviceIdentification, t, 1));
+        secretTypes.stream().map(t -> this.getUpdatedSecretsForActivation(deviceIdentification, t))
+                   .flatMap(Collection::stream).collect(collectingAndThen(toList(), this.secretRepository::saveAll));
+    }
+
+    public boolean hasNewSecret(final String deviceIdentification, final SecretType secretType) {
+        return this.secretRepository.getSecretCount(deviceIdentification, secretType, SecretStatus.NEW) > 0;
+    }
+
+    private List<DbEncryptedSecret> getUpdatedSecretsForActivation(final String deviceIdentification,
+            final SecretType secretType) {
+        List<DbEncryptedSecret> updatedSecrets = new ArrayList<>();
+        final Optional<DbEncryptedSecret> activeSecretOptional = this
+                .getSingleDbEncryptedSecret(deviceIdentification, secretType, SecretStatus.ACTIVE);
+        if (activeSecretOptional.isPresent()) {
+            final DbEncryptedSecret currentSecret = activeSecretOptional.get();
+            currentSecret.setSecretStatus(SecretStatus.EXPIRED);
+            updatedSecrets.add(currentSecret);
+        }
```

Comment:
I agree with you in general, but in this case I think it is fine: because the retrieval of the optional and the use of the optional are in the same method, close together, I think is all the same.